framework/src/headers/ip/v6/icmp/router_advertisement.rs

```diff
@@ -0,0 +1,281 @@
+                     milliseconds, between retransmitted Neighbor
+                     Solicitation messages.  Used by address resolution
+                     and the Neighbor Unreachability Detection algorithm
+                     (see Sections 7.2 and 7.3).  A value of zero means
+                     unspecified (by this router).
+
+   Possible options:
+
+      Source link-layer address
+                     The link-layer address of the interface from which
+                     the Router Advertisement is sent.  Only used on
+                     link layers that have addresses.  A router MAY omit
+                     this option in order to enable inbound load sharing
+                     across multiple link-layer addresses.
+
+      MTU            SHOULD be sent on links that have a variable MTU
+                     (as specified in the document that describes how to
+                     run IP over the particular link type).  MAY be sent
+                     on other links.
+
+      Prefix Information
+                     These options specify the prefixes that are on-link
+                     and/or are used for stateless address
+                     autoconfiguration.  A router SHOULD include all its
+                     on-link prefixes (except the link-local prefix) so
+                     that multihomed hosts have complete prefix
+                     information about on-link destinations for the
+                     links to which they attach.  If complete
+                     information is lacking, a host with multiple
+                     interfaces may not be able to choose the correct
+                     outgoing interface when sending traffic to its
+                     neighbors.
+      Future versions of this protocol may define new option types.
+      Receivers MUST silently ignore any options they do not recognize
+      and continue processing the message.
+
+*/
+
+const MANAGED_CFG_ADDR_POS: u8 = 0;
+const OTHER_CFG_POS: u8 = 1;
+
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    icmp: Icmpv6Header<T>,
+    current_hop_limit: u8,
+    reserved_flags: u8,
+    router_lifetime: u16,
+    reachable_time: u32,
+    retrans_timer: u32,
+    options: IcmpOptions,
+    _parent: PhantomData<T>,
+}
+
+impl<T> Default for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn default() -> Icmpv6RouterAdvertisement<T> {
+        Icmpv6RouterAdvertisement {
+            icmp: Icmpv6Header {
+                msg_type: IcmpMessageType::RouterAdvertisement as u8,
+                code: 0,
+                checksum: 0,
+                ..Default::default()
+            },
+            current_hop_limit: 0,
+            reserved_flags: 0,
+            router_lifetime: 0,
+            reachable_time: 0,
+            retrans_timer: 0,
+            options: IcmpOptions { options: 0 },
```

Comment:
the one thing I am not seeing is the parsing of the source link address from options.  That would be the MAC address we need, and it is possible that it is not there; however, we need to rely on it being there for router discovery.  

Since this is my first review, I maybe completely missing something, but I would expect to see on this message type a way a get_source_link_layer_address() that returns an optional

Reply:
cant we pull the mac from the macheader?

Reply:
@scrollins should be able to, unless it's part of the spec? will have to relook.

Reply:
Yea, I lose the connection from the spec to this code.  Just the part...

```
Possible options:

      Source link-layer address
                     The link-layer address of the interface from which
                     the Router Advertisement is sent.  Only used on
                     link layers that have addresses.  A router MAY omit
                     this option in order to enable inbound load sharing
                     across multiple link-layer addresses.

      MTU            SHOULD be sent on links that have a variable MTU
                     (as specified in the document that describes how to
                     run IP over the particular link type).  MAY be sent
                     on other links.

      Prefix Information
                     These options specify the prefixes that are on-link
                     and/or are used for stateless address
                     autoconfiguration.  A router SHOULD include all its
                     on-link prefixes (except the link-local prefix) so
                     that multihomed hosts have complete prefix
                     information about on-link destinations for the
                     links to which they attach.  If complete
                     information is lacking, a host with multiple
                     interfaces may not be able to choose the correct
                     outgoing interface when sending traffic to its
                     neighbors.
```

Reply:
So, I would expect to be able to `get_link_layer_address` on something.  If it is somehwere I am cool.

Also, the MTU and Prefix Information will eventually be needed when we "Know our hosts"

Reply:
@pauljamescleary yeah, it's something we can pull from the src of the ethernet/macheader if it's not in the options it seems?

Reply:
Good info - http://www.telecom.otago.ac.nz/tele301/student_html/v6bootcamp-observer-advertisements.html

Reply:
@pauljamescleary yep. Just saying that we could look at the macheader src *if* the link layer address is not there. They should be the same i reckon. Worth chekcing-in to it.

Reply:
@zeeshanlakhani interesting as the RFC doesn't say anything about it.  Just that there is this `ICMPOptions type=Source link-layer address` that is optional, and that field is what is referenced all over the doc.

I see the `options` in there, just no way to pull out the mac address from it

Reply:
Sorry, that the source link-layer address would be in the message payload

Reply:
I think we can pull it out if it is an option, otherwise look at the eth header. Would be nice to get an example to test from

Reply:
Found this...
https://www.cloudshark.org/captures/006e99788903

http://packetlife.net/captures/IPv6_NDP.cap

Interesting that you could in theory look at the eth header, but I wonder how reliable that is?  Are there circumstances where the source link-layer address would be _different_?

Reply:
according to those pcaps, the source link layer is the same as the ethernet source address. i'm not sure if that will always be the case

Reply:
i wasn't aware that we would need anything from the options fields so i intentionally left it generic. i thought the options fields were somewhat trivial being populated. i'll do some work on making sure we're parsing those fields.

Reply:
Just following up from all the other conversation, the idea is that we _should_ take the source link-layer address from the options if it is there.

If it is not there, would be nice to have logic of whether or not we need to care about this Router Advertisement.  For example, when doing auto configuration, we might not get it and could silently discard that message.

If we decide to fall-back to using the eth header, I am ok

Reply:
@pauljamescleary as far as memory layout goes, the idea is that we need to be able to peek ahead `x` number of bytes to see if we have a source-link-layer address and otherwise handle the other case. That's the work @scrollins and I discussed recently.

Reply:
That's the part I guess I'm not understanding. Just by peeking ahead, how would you know the bits you are looking at is the source link address and not something else? There's gotta be some indicator somewhere telling you what the `options` field contains right?

Reply:
if you peek ahead a certain number of bits, you can pull off information like what type of option it is, how many bytes the field is, etc. unfortunately, there are no indicators that tell you what kind of options are contained within the icmp packet.

Reply:
@drunkirishcoder - https://tools.ietf.org/html/rfc4861#section-4.6

```
Type           8-bit identifier of the type of option
```
The first byte is the type identifier.  You need to parse that in order to know what to parse next.

There is also a length so you know where the option ends

Reply:
@scrollins @pauljamescleary perfect, that makes it much easier. 

@drunkirishcoder so yeah, seems like there *should be* a format to follows. Without that, I was saying we could peek and then do our own verification. But, if things follow this format (that paul linked) and mention a length, we should be good. I should have looked further in the doc :).

Reply:
👍 that's what I was trying to find the other day but my google-fu failed me. So yeah I don't know if peek is necessary since we can parse and determine exactly what we received.

Reply:
they all have the type and length fields. i thought we still needed to peek to see if there were more bytes, no? even though the spec lists the options in a certain order, i've seen them ordered differently between two pcaps. one followed the spec while the other didnt. i'm not sure if that was another router not being a good citizen??

Reply:
i'd say adhear to the what it's supposed to be. If it's not true, default to the mac src.